extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -250,43 +290,31 @@ private boolean shouldAutoRedirect(TenantConfigContext configContext, RoutingCon
                 });
     }
 
-    private Uni<SecurityIdentity> performCodeFlow(IdentityProviderManager identityProviderManager,
-            RoutingContext context, TenantConfigContext configContext, String code) {
+    private boolean isRedirectFromProvider(RoutingContext context, TenantConfigContext configContext) {
```

Comment:
I'm not 100% sure how reliable is this. Is it really needed?

Reply:
@pedroigor Good point, I've been contemplating above about it, it is the best effort at trying to figure out how to fail the request best if the state cookie is missing.

On the `main` it is an immediate 401. But  on the `main` it is also the problem that if the `code` is detected then it is assumed it is the redirect from KC - but with the `error` and the Apple's `form_post` response mode to be supported next it just does not work - especially with `form_post` - we can't start checking the post data without asserting the state cookie exists, so this refactoring is making these checks dependent on the availability of the state cookie.

So now, if the state cookie is missing and 1)  If no `code` is there - we fail indirectly by challenging the user with a redirect to KC - this is how it works on `main`, if `code` is there - we check `Referer` and if proved it comes from KC only then do we stop everything with 401. 

So if the `Referer` check fails then instead of 401 the user will be redirected to KC to re-authenticate. And as I said above, we can follow up with adding the property what to do if the state cookie is missing, the `code` (`error`) (query or form post) present - do 401 or redirect to KC. In fact I started with the property first but then thought it was unnecessary yet.

Does it clarify it ?

 
We have a test expecting 401 in case of a missing state cookie and this code properly gives 401 as expected.

Reply:
I see. I'm still trying to figure out if this has some security implication because the header can be easily forged into the request. But don't think there is and it should improve UX.

Makes sense?

Reply:
@pedroigor Hi Pedro, 

> I'm still trying to figure out if this has some security implication because the header can be easily forged into the request. But don't think there is and it should improve UX. Makes sense?

I'm not sure I follow, can you clarify please ?

This header is only checked in the `getChallenge` method. We don't do any positive decision based on its match - we only check it to decide how to fail if `session cookie is not available but 'code' is`.

If it matches the Keycloak authorization URL then it is 401 (meaning the authentication has failed, no retries), if it does not match it - it is a new unauthenticated request, the user goes to Keycloak/etc. This referer won't be used to redirect the user.

We can totally drop this check, but that will mean it will always be only a redirect to Keycloak `if the state cookie is absent but the code is there`; the referer check it gives us an option to react to this case in a more nuanced way.
It is really about this use case, `how to fail when the code is available but the state cookie is not`.

I'm not sure, may be I will just drop this Referer check. We have a test where a state cookie is missing but the code is, and 401 is expected. Without this check I'll have to tweak the test to expect `302` to Keycloak in this case though.